userapi/src/main/java/fk/prof/userapi/cache/ZKBasedCacheInfoRegistry.java

```diff
@@ -0,0 +1,264 @@
+package fk.prof.userapi.cache;
+
+import com.codahale.metrics.Counter;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.io.BaseEncoding;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.google.protobuf.AbstractMessage;
+import com.google.protobuf.Parser;
+import fk.prof.aggregation.AggregatedProfileNamingStrategy;
+import fk.prof.metrics.Util;
+import fk.prof.userapi.proto.LoadInfoEntities.NodeLoadInfo;
+import fk.prof.userapi.proto.LoadInfoEntities.ProfileResidencyInfo;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.api.transaction.CuratorTransactionFinal;
+import org.apache.curator.framework.recipes.locks.InterProcessLock;
+import org.apache.curator.framework.recipes.locks.InterProcessMutex;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+
+import java.nio.charset.Charset;
+import java.time.LocalDateTime;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Shared storage based on zookeeper to store the mapping profile -> ip:port and maintain resource utilization counters
+ * for each node in cluster.
+ * It is used by {@link ClusterAwareCache}.
+ *
+ * zookeeper node structure:
+ * /fkprof-userapi                                                                    (session namespace)
+ * /fkprof-userapi/nodesInfo/{ip:port} -> {@link NodeLoadInfo}                                (ephemeral)
+ * /fkprof-userapi/profilesLoadStatus/{profileName} -> {@link ProfileResidencyInfo}           (ephemeral)
+ *
+ * For resource utilization counter, currently only cached profile count is stored
+ *
+ * @see NodeLoadInfo
+ * @see ProfileResidencyInfo
+ *
+ * Created by gaurav.ashok on 01/08/17.
+ */
+class ZKBasedCacheInfoRegistry implements CacheInfoRegistry {
+    private static final Logger logger = LoggerFactory.getLogger(ZKBasedCacheInfoRegistry.class);
+    private static final String distributedLockPath = "/global_mutex";
+
+    private final String zkNodesLoadInfoPath;
+    private final CuratorFramework curatorClient;
+    private final InterProcessLock interProcessLock;
+
+    private final AtomicReference<ConnectionState> connectionState;
+    private final AtomicReference<LocalDateTime> lastZkLostTime;
+    private final AtomicBoolean recentlyZkConnectionLost;
```

Comment:
If the idea is just to have immediate visibility across threads, then making these volatile will also suffice rather than atomic